GO: Alternate fix for race conditions

Snarling opened the pull request:
@d0sboots @ficocelliguy I believe this stops any possibility for there to be multiple AI turns pending at the same time, but let me know what you think and if there's some other potential race condition that doesn't seem to be addressed.

Because this PR includes a location change + function name change + changes to the function, it makes more sense to look at the individual race condition 1 and 2 commits for reviewing those specific things, instead of the overall changes for the PR which are hard to see the changes in.

The move and rename commits had no changes other than location and name of the function.

6d01457: Moved getAIMove to a different location since it's not netscript specific. No substantive changes, just a location change.

48bfb2a: Close a weird 300ms gap where it's actually the player's turn because the AI move was already applied to the board, but the Go.nextTurn promise for the previous AI turn has not yet resolved. There was an extra sleep that was causing this weird mismatched in-between time. I slightly increased one of the existing sleeps to compensate removing the extra sleep.

afeb6e7: More directly deals with the possibility of multiple concurrent AI turns due to being triggered from different sources. Now AI turns are disallowed from being triggered until the previous one has resolved. If multiple are triggered simultaneously then they will both return the same turn, instead of being two separate instance of the AI attempting to take a turn.

708ba27: Renamed getAITMove to makeAIMove since it actually triggers the AI to make the move. No substantive changes, just a name change.

3eb4dc7: Removed the functionality from the UI where loading Go will take the AI's turn if a game was loaded where it was already the AI's turn. Instead this is handled by the game automatically on game load. Also un-stated the waitingOnAI variable on the gameboardwrapper, because it's a variable handled by Go internal game state and not React state.

Snarling committed:
Move getAIMove function to goAI file

Function was being used by both the UI and the API, it did not belong in netscriptGoImplementation

Snarling committed:
Address race condition 1

* Remove extra sleep after the AI move has already been applied to the board and before resolving nextTurn. Prevents issue where the boardState looks like it's the player's turn, but the nextTurn promise has not been resolved yet.
* Increased initial sleep (between finding move and applying move) from 400ms to 500ms, to account for removing the other sleep entirely.
* Combined two instances of emitting a go event and resolving with the result, this will make the next commit easier.

Snarling committed:
Address race condition 2

Don't allow multiple AI moves to be queued at the same time (needed due to both API and UI being able to directly trigger AI moves).

Snarling committed:
Rename function

Renamed getAIMove to makeAIMove, to make it clear that it is not just for getting the move (like the getMove function) but is for actually applying the move to the board as well.
Not really necessary in this PR, just an editorial change since I was messing with that function anyway.

Snarling committed:
Handle missed AI turn on game load

* If it's the AI's turn on game load, take the AI turn after loading
* Remove AI turn handling from UI pageload
* Remove React state variable for whether it's the AI's turn

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -21,13 +21,23 @@ import { findAnyMatchedPatterns } from "./patternMatching";
-    resolve = res;
+    resolve = (play) => {
+      // Once the promise resolves, stop blocking new AI moves
+      isAIBusy = false;
```
Subtle point: I think you should set this in a `finally` chained off of `getMove`. I don't *think* it makes a difference now, but it would if any of these promises rejected.

Snarling replied:
Good idea, implemented in a471108 / 24cd914

d0sboots commented:
Great, this is precisely the kind of race-condition avoidance I was thinking of! This will also make it easier to make a change I have vaguely kicking around in my head, which is to move Go AI processing to a Web Worker.

Snarling committed:
Remove async wrapper for main makeAIMove

makeAIMove returns a promise directly without awaiting, it doesn't need to be async

Snarling committed:
Rename variable

Resulting play from getMove renamed from `result` to `play` to make it more clear what it is

Snarling committed:
Make suggested change

Clearing isAIBusy is handled in a finally block
Right now the Go.nextTurn promise has no way to ever error out. (and If the getMove promise errors out, the Go.nextTurn promise simply never resolves. Will address that in the next commit.

Snarling committed:
Unwrap Go.nextTurn

Instead of Go.nextTurn being a manually-resolving promise, it is a chained promise directly off getMove. This allows any errors from getMove to reject the nextTurn promise directly instead of also needing a manually-triggered rejection function.

Snarling committed:
Trigger GoEvent always at end of AI turn

Snarling committed:
Avoid passing on new board if old board has been replaced

If it's a stale game by the time getMove is finished (player started a new game while AI was thinking), and the AI's move on the old board would have been a pass, the old logic would have had the AI pass on the new board.

Snarling committed:
Move getPreviousMove to a different file

Next commit will make getPreviousMove be used outside of just netscript, so it didn't belong in the netscript-specific file anymore

Snarling committed:
nextTurn is no longer possibly null

Since the API was already initializing it on demand, the null value was never being used.

Now the API no longer initializes it, it is initialized on game load.

Also slightly adjusted the way the AI busy tracking works, since other things can change Go.nextTurn (endGoGame). endGoGame should never overwrite Go.nextTurn while the AI is still busy taking their turn, but it seems slightly more robust this way.

Snarling committed:
More specific definition for Play type + remove asyncs

* Moves always have numeric x/y, pass and gameover have null x/y
* Removed some extra async wrappers from netscriptGoImplementation, where promises were being returned directly without any awaiting

Snarling committed:
Partial revert

Previous commit caused some weird test errors that seem to be due to an incomplete test environment, but I don't feel like fixing it so the extra async wrappers are back

Snarling committed:
Fix test?

This is a reversion, should have reverted this in previous commit.

Snarling commented:
Notable further changes:

* Go.nextTurn is no longer possibly null. It was being initialized as soon as it was accessed anyway and the only null case was on initial game load, so now it is just initialized on load (15ff42a)
* There was a minor bug where if the AI was trying to pass, and the player started a new game before the AI finished determining the next move, the pass would be applied to the current game. That was addressed in b0bbfe7

There is a bug currently where passTurn is being called twice but I'm not sure the best way to fix it. Right now it's being called during getMove (which is not supposed to be actually changing the state of the game), and then called again during makeAIMove. I'd like to remove it from getMove. That'll probably be the last thing included in this PR once I figure out a good way to do that.

Snarling committed:
Only pass once

If the AI's move is a pass, only call passTurn once, insice makeAIMove instead of getMove

Snarling commented:
Alright, I think I'm done messing around with this. @d0sboots @ficocelliguy FYI in case you want to look over / give feedback on the changes before I merge them

d0sboots commented on src/Go/SaveLoad.ts:
```diff
@@ -77,6 +78,18 @@ export function loadGo(data: unknown): boolean {
   Go.stats = stats;
+
+  // If it's the AI's turn, initiate their turn, which will populate nextTurn
+  if (currentGame.previousPlayer === GoColor.black && currentGame.ai !== GoOpponent.none) makeAIMove(currentGame);
```
IMO all-on-one-line style for if statements is a bit shady (it's lead to various high-profile bugs in other applications), but especially if you have an else block you should always use curly braces for the if block.

d0sboots reviewed (approved):
Nice, "Remove async wrapper for main makeAIMove" and "Unwrap Go.nextTurn" were things I was looking at too, but didn't comment on because I didn't want to expand this PR unduly. :D
